circuits/types/field/src/from_bits.rs

```diff
@@ -31,7 +32,45 @@ impl<E: Environment> FromBits for Field<E> {
             // Check if all excess bits are zero.
             let should_be_zero = bits_le[size_in_bits..].iter().fold(Boolean::constant(false), |acc, bit| acc | bit);
             // Ensure `should_be_zero` is zero.
-            E::assert_eq(E::zero(), should_be_zero);
+            // TODO: `E::assert_eq` does not enforce constant constraints.
+            if should_be_zero.is_constant() && should_be_zero.eject_value() {
+                E::halt("The excess bits are not zero.")
+            } else {
+                E::assert_eq(E::zero(), should_be_zero)
+            }
+        }
+
+        // If the number of bits is greater than `size_in_data_bits`, then check that it is a valid field element.
+        if num_bits > size_in_data_bits {
```

Comment:
If we have an `is_less_than` and `is_less_than_equal_to`, how come we're rewriting the logic here?

Reply:
`Field::is_less_than` is called on `Field` circuits. 
`Field::is_less_than` invokes `to_bits` which produces 254 constraints.
Since we already have `bits_le`, we shouldn't pay for redundant computation.

Reply:
Might be worth extracting out a common helper function to avoid duplication.

Reply:
I would avoid the helper functions altogether. It was a big factor in contributing to `gadgets/`' readability issues